repl/repl.go

```diff
@@ -78,29 +84,77 @@ func EvalString(what string) (res string, errs []string, formatted string) {
 	return
 }
 
-func Interactive(in io.Reader, out io.Writer, options Options) {
+func Interactive(options Options) int {
 	options.NilAndErr = true
 	s := eval.NewState()
 	macroState := object.NewMacroEnvironment()
 
-	scanner := bufio.NewScanner(in)
 	prev := ""
-	prompt := PROMPT
+
+	term, err := terminal.Open()
+	if err != nil {
+		return log.FErrf("Error creating readline: %v", err)
+	}
+	defer term.Close()
+	term.LoggerSetup()
+	term.SetPrompt(PROMPT)
+	options.Compact = true // because terminal doesn't do well will multi-line commands.
+	term.NewHistory(options.MaxHistory)
+	_ = term.SetHistoryFile(options.HistoryFile)
+	// Regular expression for "!nn" to run history command nn.
+	historyRegex := regexp.MustCompile(`^!(\d+)$`)
 	for {
-		fmt.Fprint(out, prompt)
-		scanned := scanner.Scan()
-		if !scanned {
-			return
+		rd, err := term.ReadLine()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// To avoid trailing prompt due to prompt refresh on the log output
+				// that's a bit ugly that it's necessary. consider handling in terminal.Close
+				term.SetPrompt("")
+				log.Infof("Exit requested") // Don't say EOF as ^C comes through as EOF as well.
+				return 0
+			}
+			return log.FErrf("Error reading line: %v", err)
+		}
```

Comment:
```suggestion
		if errors.Is(err, io.EOF) {
			// To avoid trailing prompt due to prompt refresh on the log output
			// that's a bit ugly that it's necessary. consider handling in terminal.Close
			term.SetPrompt("")
			log.Infof("Exit requested") // Don't say EOF as ^C comes through as EOF as well.
			return 0
		}
		if err != nil {
			return log.FErrf("Error reading line: %v", err)
		}
```

Reply:
thanks for the reminder, that code moved to terminal/ so isn't needed at all anymore